### ISSUE TITLE
Autocomplete error?

### ISSUE THREAD

**Issue by JVMV** [godotengine/godot#83080]
### Godot version

4.2 dev-6

### System information

Windows 11, forward+

### Issue description

![image](https://github.com/godotengine/godot/assets/51789223/18721853-a2a0-471c-8df4-710f2c53c46b)

essentially the engine is assuming a different type from what i cast the variable as, based on the return type of the function

1) it's not recognizing that im placing the wrong thing into the variable
2) it's not respecting my cast of the variable and instead is just assuming it's the return type

<NavigationPolygon#-9223372003619109625>
[(0, 0)]

--- Debugging process stopped ---
  res://Scripts/Enemies/BouncyEye.gd:48 - Parse Error: Cannot find property "pick_random" on base "PackedVector2Array".
  res://Scripts/Enemies/BouncyEye.gd:48 - Parse Error: Function "pick_random()" not found in base PackedVector2Array.

### Steps to reproduce

Open the reproduction project.

### Minimal reproduction project

[ReproductionArray.zip](https://github.com/godotengine/godot/files/12851614/ReproductionArray.zip)

**Comment by joshua-holmes** [comment]
So from a user level, I think the solution is to declare `decomposed` like this instead:
```
var decomposed: Array[PackedVector2Array]
```
This is because `Geometry2D::decompose_polygon_in_convex` returns an array of `PackedVector2Array`, not a `PackedVector2Array` itself. [documentation link](https://docs.godotengine.org/en/stable/classes/class_geometry2d.html#class-geometry2d-method-decompose-polygon-in-convex)

When I declare that variable with the correct type, I do not get that error and I am allowed to call `pick_random()`, as I should be able to.

However, Godot's perspective, the engine *should* prevent you from assigning a value of the wrong data type. For example, this is not allowed and Godot will stop you:
```
var i: int
i = "hi"
```
But it doesn't stop you from doing what you are doing in your example. Just as a test, it also doesn't prevent me from doing this:
```
var navpolygonobj: NavigationPolygon = NavigationPolygon.new()
var navpoly_poly: PackedVector2Array = navpolygonobj.polygons

var decomposed: PackedVector2Array
decomposed = [navpoly_poly]
```
Notice I place `navpoly_poly` in an `Array`, making it a `Array[PackedVector2Array]` and it didn't give me any errors.

**Comment by joshua-holmes** [comment]
All that being said, @Calinou, would you be able to assign this ticket to me? :grin: I would like to see if I can find a solution

**Comment by KoBeWi** [comment]
Seems fixed in 4.3
![image](https://github.com/godotengine/godot/assets/2223172/a4868bbe-985a-4877-b0fe-29908faa6145)
